### PR TITLE
Major project refactor and clean-up

### PR DESCRIPTION
**Pull request by manusa**
- Created the base MockServer interface that now exposes most of the underlying MockWebServer methods required for test assertions
- Moved downstream `getLastRequest` method to MockServer + added cache behavior
- Updated dependencies
- Refactored code and removed bugs + smells
- CRUD dispatcher now handles PUT independently of POST
- Added more test coverage

/cc @shawkins Most of the things needed downstream have been exposed. I'm not sure if I may be leaving something out.

### CODE REVIEW

**Comment by manusa** [comment]
> @manusa I didn't have a particular need beyond exposing the namespace attribute. The only other thing I can think of that the mock can't handle, which can be handled separately of course, is that repeated calls to the same websocket url don't work - that is once you upgrade to a websocket and it gets closed the underlying executor gets closed and can't be used for a subsequent call to the same url.

Do you have an example for this?

I'm assuming you might be talking about this part:
https://github.com/fabric8io/mockwebserver/blob/6e9eb1bebb1c20594aca6b571f4fa7e42f9a2d6c/src/main/java/io/fabric8/mockwebserver/internal/WebSocketSession.java#L160-L172

File: src/main/java/io/fabric8/mockwebserver/MockServerException.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.mockwebserver;
+
+public class MockServerException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (manusa) [reply src/main/java/io/fabric8/mockwebserver/MockServerException.java:18]: It's a good suggestion. I'll add it to the current PR.

**Comment by manusa** [comment]
Hi @shawkins 
The last commit should take care of your scenario. Could you please provide your review?

**Review by shawkins — APPROVED** [review]
LGTM

**Pull request merged**